c/authService.c

```diff
@@ -173,16 +196,223 @@ static int serveAuthCheck(HttpService *service, HttpResponse *res) {
     respondWithError(res, HTTP_STATUS_BAD_REQUEST, "Unexpected access level");
     return 0;
   }
-  /* printf("query: user %s, class %s, entity %s, access %d\n", userName, class,
-      entity, access); */
+
   privilegedServerName = getConfiguredProperty(service->server,
       HTTP_SERVER_PRIVILEGED_SERVER_PROPERTY);
   rc = zisCheckEntity(privilegedServerName, userName, class, entity, access,
       &reqStatus);
+
   respond(res, rc, &reqStatus);
   return 0;
 }
 
+int serveAuthCheckByParams(HttpService *service, char *userName, char *class, char *entity, int access, JsonObject *envSettings) {
+  int rc = 0;
+  int rbacParm;
+  Json *rbacObj = jsonObjectGetPropertyValue(envSettings, "ZWED_dataserviceAuthentication_rbac");
+  if (rbacObj == NULL) { // Env variable doesn't exist, so check server config
+    JsonObject *dataserviceAuth = jsonObjectGetObject(service->server->sharedServiceMem, "dataserviceAuthentication");
+    rbacParm = jsonObjectGetBoolean(dataserviceAuth, "rbac");
+  } else { // Use env variable value
+    rbacParm = jsonAsBoolean(rbacObj);
+  }
+  
+  CrossMemoryServerName *privilegedServerName = getConfiguredProperty(service->server,
+      HTTP_SERVER_PRIVILEGED_SERVER_PROPERTY);
+  ZISAuthServiceStatus reqStatus = {0};
+  if (!rbacParm) {
+    return rc; // When rbac isn't enabled, we don't try to check the auth query
+  }
+  rc = zisCheckEntity(privilegedServerName, userName, class, entity, access,
+      &reqStatus);
+  zowelog(NULL, LOG_COMP_ID_SECURITY, ZOWE_LOG_DEBUG2,
+           "RBAC check occurred for entity '%s' class '%s' access '%d' , rc: %d", entity, class, access, rc);
+  return rc;
+}
+
+int getProfileNameFromRequest(char *profileName, StringList *parsedFile, char *method, int instanceID, HttpResponse *response) {
+  char type[STRING_BUFFER_SIZE]; // core || config || service
+  char productCode[STRING_BUFFER_SIZE];
+  char rootServiceName[STRING_BUFFER_SIZE];
+  char subUrl[SAF_SUB_URL_SIZE][STRING_BUFFER_SIZE] = {0};
+  char scope[STRING_BUFFER_SIZE];
+  char pluginID[STRING_BUFFER_SIZE]; 
+  char serviceName[STRING_BUFFER_SIZE];
+  char urlSegment[STRING_BUFFER_SIZE];
+  int subUrlIndex = 0;
+  
+  snprintf(urlSegment, STRING_BUFFER_SIZE, "%s", stringListPrint(parsedFile, 1, 1, "/", 0));
+  StringListElt *pathSegment = firstStringListElt(parsedFile);
+
+  strupcase(urlSegment);
+  if (instanceID < 0) { // Set instanceID
+    instanceID = 0;
+  }
+  if (strcmp(urlSegment, "PLUGINS") != 0) {
+    zowelog(NULL, LOG_COMP_ID_SECURITY, ZOWE_LOG_DEBUG2,
+           "parsedFile urlSegment check didn't match.");
+    subUrlIndex = -1;
+    while (pathSegment != NULL) {
+      snprintf(urlSegment, STRING_BUFFER_SIZE, "%s", pathSegment->string);
+      strupcase(urlSegment);
+      if (rootServiceName == NULL)
+      {
+        snprintf(rootServiceName, STRING_BUFFER_SIZE, urlSegment);
+      } else {
+        if (subUrlIndex < SAF_SUB_URL_SIZE) {
```

Comment:
Here is a link to a Zowe doc regarding RBAC query best practices: https://docs.zowe.org/stable/extend/extend-desktop/mvd-dataservices/#using-dataservices-with-rbac

Currently there's no hard limit on URL subsections, but there is a hard char limit (after when reached, we just trim instead of rejecting) so in all likelihood it's unlikely for there to be more than 16 (SAF_SUB_URL_SIZE) subsections but I can bump it up to 32 to avoid any problems. When it's larger than SAF_SUB_URL_SIZE, we don't need to handle this case because we just stop adding URL subsections to profile name

Reply:
I'll add a comment here as well to make that more clear